lib/spack/spack/package.py

```diff
@@ -1147,6 +1152,10 @@ def fetcher(self):
     @fetcher.setter
     def fetcher(self, f):
         self._fetcher = f
+        if isinstance(f, StageComposite):
+            self._fetcher[0].package = self
+        else:
+            self._fetcher.package = self
```

Comment:
I'm not sure polymorphism works in this case.  The package is only being set so that the `GitFetchStrategy` can call back on it to get submodules needed for the concretized spec.  The submodules returned by `Package.submodules()` are valid only for the package's source and not each of the package's resources.  So, the package should only be set for the root fetcher.  The designation of "root" fetcher is arbitrary and is only known by the package, this the decision tree to set the fetcher's package is in the package's code, not the fetchers.

But, there is a typo in this code.  It should read `if isinstance(f, fs.FetchStrategyComposite)`.